raft/log.go

```diff
@@ -1145,9 +1270,27 @@ func (l *Log) waitUncommitted(index uint64) error {
 }
 
 // append adds a log entry to the list of entries.
-func (l *Log) append(e *LogEntry) {
-	//l.tracef("append: idx=%d, prev=%d", e.Index, l.lastLogIndex)
-	assert(e.Index == l.lastLogIndex+1, "non-contiguous log index(%d): idx=%d, prev=%d", l.id, e.Index, l.lastLogIndex)
+func (l *Log) append(e *LogEntry) error {
+	// Exit if log is not in a running state.
+	// Ignore replayed entries.
+	if l.state == Stopped {
+		return ErrClosed
+	} else if e.Index <= l.lastLogIndex {
```

Comment:
Yes, since the applied index was moved to the FSM entirely, there's a small window where we can get duplicate indices immediately following a snapshot. We grab the applied index before making  the snapshot request from the FSM to ensure that we have all entries.